pallets/loans-ref/src/types.rs

```diff
@@ -0,0 +1,662 @@
+#[derive(Encode, Decode, Clone, Default, TypeInfo, MaxEncodedLen)]
+pub struct PortfolioValuation<Balance> {
+	// Computed portfolio valuation for the given pool
+	value: Balance,
+
+	// Last time when the portfolio valuation was calculated for the entire pool
+	last_updated: Moment,
+}
+
+impl<Balance> PortfolioValuation<Balance>
+where
+	Balance: tokens::Balance,
+{
+	pub fn new(value: Balance, when: Moment) -> Self {
+		Self {
+			value,
+			last_updated: when,
+		}
+	}
+
+	pub fn value(&self) -> Balance {
+		self.value
+	}
+
+	pub fn last_updated(&self) -> Moment {
+		self.last_updated
+	}
+
+	pub fn update_with_pv_diff(
+		&mut self,
+		old_pv: Balance,
+		new_pv: Balance,
+	) -> Result<(), ArithmeticError> {
+		match new_pv.cmp(&old_pv) {
+			Ordering::Greater => self.value.ensure_add_assign(new_pv.ensure_sub(old_pv)?),
+			Ordering::Less => self.value.ensure_sub_assign(old_pv.ensure_sub(new_pv)?),
+			Ordering::Equal => Ok(()),
+		}
+	}
+}
+
+/// Information about how the portfolio valuation was updated
+#[derive(Encode, Decode, Clone, PartialEq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub enum PortfolioValuationUpdateType {
+	/// Portfolio Valuation was fully recomputed to an exact value
+	Exact,
+	/// Portfolio Valuation was updated inexactly based on loan status changes
+	Inexact,
+}
+
+/// The data structure for storing a specific write off policy
+#[derive(Encode, Decode, Clone, PartialEq, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct WriteOffState<Rate> {
+	/// Number in days after the maturity has passed at which this write off policy is valid
+	overdue_days: u32,
+
+	/// Percentage of present value we are going to write off on a loan
+	pub percentage: Rate,
+
+	/// Additional interest that accrues on the written off loan as penalty
+	pub penalty: Rate,
+}
+
+impl<Rate> WriteOffState<Rate>
+where
+	Rate: FixedPointNumber,
+{
+	fn is_not_overdue(&self, maturity_date: Moment, now: Moment) -> Result<bool, ArithmeticError> {
+		let overdue_secs = SECONDS_PER_DAY.ensure_mul(self.overdue_days.ensure_into()?)?;
+		Ok(now >= maturity_date.ensure_add(overdue_secs)?)
+	}
+
+	pub fn find_best(
+		policy: impl Iterator<Item = WriteOffState<Rate>>,
+		maturity_date: Moment,
+		now: Moment,
+	) -> Option<WriteOffState<Rate>> {
+		policy
+			.filter_map(|p| p.is_not_overdue(maturity_date, now).ok()?.then_some(p))
+			.max_by(|a, b| a.overdue_days.cmp(&b.overdue_days))
+	}
+
+	pub fn status(&self) -> WriteOffStatus<Rate> {
+		WriteOffStatus {
+			percentage: self.percentage,
+			penalty: self.penalty,
+		}
+	}
+}
+
+/// Diferent kinds of write off status that a loan can be
+#[derive(Encode, Decode, Clone, PartialEq, Default, TypeInfo, RuntimeDebug, MaxEncodedLen)]
+pub struct WriteOffStatus<Rate> {
+	/// Percentage of present value we are going to write off on a loan
+	pub percentage: Rate,
+
+	/// Additional interest that accrues on the written down loan as penalty per sec
+	pub penalty: Rate,
+}
```

Comment:
Where are the variants?
If we no longer need the `index` of a write-off group and instead of that we store the `percentage` and `penalty`, we can assume 0% for `None` and `100%` for `WriteOff`, and all the variants can be simplified to a struct. This reduces a lot of complexity when the status is read/written later.